interactive_engine/executor/ir/graph_proxy/src/adapters/gs_store/read_graph.rs

```diff
@@ -94,19 +103,39 @@ where
                 })
                 .unwrap_or(DEFAULT_SNAPSHOT_ID);
             let label_ids = encode_storage_labels(params.labels.as_ref())?;
-            let prop_ids = encode_storage_prop_keys(params.columns.as_ref())?;
-            let filter = params.filter.clone();
+            let row_filter = params.filter.clone();
+
+            let (condition, row_filter_exists_but_not_pushdown) =
+                encode_storage_row_filter_condition(row_filter.as_ref(), self.row_filter_pushdown);
+
+            // props that will be returned by storage layer
+            let prop_ids = if self.column_filter_pushdown {
+                if row_filter_exists_but_not_pushdown {
+                    // need to call filter_limit!, so get columns in row_filter and params.columns
+                    extract_needed_columns(row_filter.as_ref(), params.columns.as_ref())?
```

Comment:
To avoid encode_storage_prop_keys() twice for both `prop_ids` and `further_prop_ids`, we can 
```
// properties that need to be cached and will be used in further compute steps
let cache_prop_ids = encode_storage_prop_keys(params.columns.as_ref())?;

let prop_ids = if self.column_filter_pushdown {
     if row_filter_exists_but_not_pushdown {
            / / need to call filter_limit!, so get columns in row_filter and params.columns
            extract_needed_columns(row_filter.as_ref(), cache_prop_ids)?
    } else {
           cache_prop_ids.clone()
    }
...
}

Reply:
added